docs/rfcs/020-pageserver-s3-coordination.md

```diff
@@ -0,0 +1,75 @@
+
+In this mode second pageserver needs to have the same picture of s3 files to be able to load layers on-demand. To accomplish that second pageserver cannot run gc/compaction jobs. Instead it needs to receive updates for index contents. (There is no need to run walreceiver on the second pageserver then).
+
+## Observations
+
+- If both pageservers ingest wal then their layer set diverges, because layer file generation is not deterministic
+- If one of the pageservers does not ingest wal (and just picks up layer updates) then it lags behind and cannot really answer queries in the same pace as the primary one
+- Can compaction help make layers deterministic? E g we do not upload level zero layers and construction of higher levels should be deterministic. This way we can guarantee that layer creation by timeout wont mess things up. This way one pageserver uploads data and second one can just ingest it. But we still need some form of election
+
+## Solutions
+
+### Manual orchestration
+
+One possible solution for relocation case is to orchestrate background jobs from outside. The oracle who runs migration can turn off background jobs on PS1 before migration and then run migration -> enable them on PS2. The problem comes if migration fails. In this case in order to resume background jobs oracle needs to guarantee that PS2 doesnt run background jobs and if it doesnt respond then PS1 is stuck unable to run compaction/gc. This cannot be solved without human ensuring that no upload from PS2 can happen. In order to be able to resolve this automatically CAS is required on S3 side so pageserver can avoid overwriting index part if it is no longer the leading one
+
+Note that flag that disables background jobs needs to be persistent, because otherwise pageserver restart will clean it
+
+### Avoid index_part.json
+
+Index part consists of two parts, list of layers and metadata. List of layers can be easily obtained by \`ListObjects\` S3 API method. But what to do with metadata? Create metadata instance for each checkpoint and add some counter to the file name?
+
+Back to potentially long s3 ls.
+
+### Coordination based approach
+
+Do it like safekeepers chose leader for WAL upload. Ping each other and decide based on some heuristics e g smallest node id. During relocation PS1 sends "resign" ping message so others can start election without waiting for a timeout.
+
+This still leaves metadata question open and non deterministic layers are a problem as well
+
+### Avoid metadata file
+
+One way to eliminate metadata file is to store it in layer files under some special key. This may resonate with intention to keep all relation sizes in some special segment to avoid initial download during size calculation. Maybe with that we can even store pre calculated value.
+
+As a downside each checkpoint gets 512 bytes larger.
+
+If we entirely avoid metadata file this opens up many approaches
```

Comment:
I'm not sure how to do that properly. We need `disk_consistent_lsn` and `latest_gc_cutoff_lsn` to be updated periodically on the local and remote side.
All other properties are more or less static and less important, except for "metadata version" field.


While the `disk_consistent_lsn` is possible to embed into the layer name, presumably, since the layer files are being flushed only once and cause `disk_consistent_lsn` to be increased due to that.
But `latest_gc_cutoff_lsn` changes unpredictably and does not produce new layers: instead, it removes the old ones and demands that the `latest_gc_cutoff_lsn` for the new layer set is persisted (if we use the older, lesser value on the new layer set there might be issues).
How to store that without `metadata`?

More complex issues are related to "metadata version" since certain PS versions won't be compatible with certain layer formats and we need a good way to detect that before downloading layers to disk and checking their bytes.
Might be better to work on layer format more before seriously discussing this.

Generally, the biggest questions is how to access this data with the new approach? If we start to store layer size inside of each layer and we will have, say, 1000 new layers on a remote storage, what would be the good way to collect the size information from every layer without downloading them?

Reply:
> Generally, the biggest questions is how to access this data with the new approach? If we start to store layer size inside of each layer and we will have, say, 1000 new layers on a remote storage, what would be the good way to collect the size information from every layer without downloading them?

This is the same problem as with relation sizes that we need to collect on startup to populate logical size. If we go with s3 ls instead of index part we can pick layer created by latest checkpoint and extract this data from it

> How to store that without metadata?

As a kv pair in layer created on each checkpoint for example

> and we need a good way to detect that before downloading layers to disk and checking their bytes.

We can have a header in each layer file with such metadata (like it is for usual metadata) and start download sort of speculatively and abort it if we detected something unexpected.

Reply:
Oh, that looks very much similar to the homegrown compression approach with blob headers and their streaming downloads we've made, yet now it's per layer, not per timeline.

If it's indeed possible to stream a small header part of the layer, to get that kv data, we're automatically getting its S3 metadata (sha, file size, custom metadata strings that we set additionally).

-------------
I wonder, would throttling be an issue or not in such case?

https://aws.amazon.com/premiumsupport/knowledge-center/s3-request-limit-avoid-throttling/ says that 

> You can send [3,500 PUT/COPY/POST/DELETE or 5,500 GET/HEAD requests](https://docs.aws.amazon.com/AmazonS3/latest/dev/optimizing-performance.html) per second per prefix in an Amazon S3 bucket. There are no limits to the number of prefixes that you can have in your bucket.

I have failed to find the numbers for AWS IAM we're using to avoid storing actual credentials, but that was hundreds requests per second, before it started failing.

I've checked ps-2 staging and there are tenants with timelines that have 178 layers, with `mean` ~13 layers /timeline, but I'm afraid that's due to mostly inactive projects.

I like the integrity of the approach (if we're truly able to stream the layer header fast), but concerned about AWS limitations we might hit.

Reply:
> I wonder, would throttling be an issue or not in such case?

Shouldnt be much different from current usage pattern. If we can determine latest layer from `s3 ls` output then we download only one file with all the metadata

I think throttling will anyway become a problem with ondemand download/eviction being in place